taichi/transforms/constant_fold.cpp

```diff
@@ -11,61 +14,161 @@ class ConstantFold : public BasicStmtVisitor {
   ConstantFold() : BasicStmtVisitor() {
   }
 
-  void visit(UnaryOpStmt *stmt) override {
-    if (stmt->width() == 1 && stmt->op_type == UnaryOpType::cast_value &&
-        stmt->operand->is<ConstStmt>()) {
-      auto input = stmt->operand->as<ConstStmt>()->val[0];
-      auto src_type = stmt->operand->ret_type.data_type;
-      auto dst_type = stmt->ret_type.data_type;
-      TypedConstant new_constant(dst_type);
-      bool success = false;
-      if (src_type == DataType::f32) {
-        auto v = input.val_float32();
-        if (dst_type == DataType::i32) {
-          new_constant.val_i32 = int32(v);
-          success = true;
-        }
-      } else if (src_type == DataType::i32) {
-        auto v = input.val_int32();
-        if (dst_type == DataType::f32) {
-          new_constant.val_f32 = float32(v);
-          success = true;
+  struct JITEvaluatorId
+  {
+    int op;
+    DataType ret, lhs, rhs;
+    bool is_binary;
+
+    explicit operator int() const // make STL map happy
+    {
+      return (int)op | (int)!!is_binary << 7
+        | (int) ret << 8 | (int) lhs << 16 | (int) rhs << 24;
+    }
```

Comment:
Maybe we should use `JITEvaluatorId` itself as the key of the JIT evaluation kernel cache. You'll have to define a few more members for hashing and comparing two `JITEvaluatorId`, but that will make the system more robust.

Reply:
I used to have this, but I give up because, it requires a `hash` structure that knows all member of `JITEvaluatorId`, while I want to keep `JITEvaluatorId` private in this file (instead of in header).

Reply:
Sounds good. I think we are good for now, but keep in mind that if `lhs`, `rhs`, `ret` grows bigger than 256, then this hash value is no longer unique.

Reply:
`!!is_binary` prevents the situation that `is_binary` is none of 0 or 1, e.g. `is_binary = 2` will cause a serious problem here.

Reply:
I think `(int)is_binary` can only have a value of 0 or 1.
https://stackoverflow.com/questions/2725044/can-i-assume-booltrue-int1-for-any-c-compiler

Reply:
BTW, if `rhs` is at least 128 (not 256), the result of `(int) rhs << 24` can be undefined (although I believe it's correct in most compilers)
https://stackoverflow.com/questions/7622/are-the-shift-operators-arithmetic-or-logical-in-c

Reply:
How about `using JITEvaluatorIdType = uint32_t`?

Reply:
Sounds good.